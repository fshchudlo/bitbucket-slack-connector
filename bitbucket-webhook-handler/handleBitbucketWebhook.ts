import * as useCases from "./use-cases";
import { SlackAPIAdapter } from "./ports/SlackAPIAdapter";
import { BitbucketAPIAdapter } from "./ports/BitbucketAPIAdapter";
import { BitbucketNotification } from "../bitbucket-payload-types";
import { WebhookHandlerConfig } from "./webhookHandlerConfig";
import { provisionPullRequestChannel } from "./provisionPullRequestChannel";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketAPI: BitbucketAPIAdapter, config: WebhookHandlerConfig) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, bitbucketAPI, payload, config);

    switch (eventKey) {
        case "pr:opened":
            await useCases.inviteParticipantsAndSetChannelBookmark(payload, slackAPI, config.defaultChannelParticipants, channelInfo.id);
            await useCases.tryBroadcastMessageAboutOpenedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        case "pr:modified":
            await useCases.sendMessageAboutPRModification(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:updated":
            await useCases.updateChannelMembers(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await useCases.sendMessageAboutReviewerAction(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:added":
            await useCases.sendMessageAboutAddedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:edited":
            await useCases.sendMessageAboutEditedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:deleted":
            await useCases.sendMessageAboutDeletedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:from_ref_updated":
            await useCases.sendMessageAboutNewCommit(payload, slackAPI, bitbucketAPI, channelInfo.id);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await useCases.sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo.id);
            await useCases.tryBroadcastMessageAboutClosedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

