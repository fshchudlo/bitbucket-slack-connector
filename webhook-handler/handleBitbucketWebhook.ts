import {
    sendCompletionMessageAndCloseTheChannel,
    setChannelTopicAndInviteParticipants,
    sendMessageAboutAddedComment,
    updateChannelMembers,
    sendMessageAboutNewCommit,
    sendMessageAboutReviewerAction,
    sendMessageAboutDeletedComment, sendMessageAboutEditedComment
} from "./use-cases";
import { SlackAPIAdapter } from "./SlackAPIAdapter";
import { BitbucketGateway } from "./BitbucketGateway";
import { sendMessageAboutPRModification } from "./use-cases/sendMessageAboutPRModification";
import { BitbucketNotification } from "../typings";
import { buildChannelName } from "./slack-helpers";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketGateway: BitbucketGateway, usePrivateChannels: boolean = true) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, payload, usePrivateChannels);
    switch (eventKey) {
        case "pr:opened":
            await setChannelTopicAndInviteParticipants(payload, slackAPI, channelInfo);
            break;
        case "pr:modified":
            await sendMessageAboutPRModification(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:updated":
            await updateChannelMembers(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await sendMessageAboutReviewerAction(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:added":
            await sendMessageAboutAddedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:edited":
            await sendMessageAboutEditedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:deleted":
            await sendMessageAboutDeletedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:from_ref_updated":
            await sendMessageAboutNewCommit(payload, slackAPI, bitbucketGateway, channelInfo);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo);
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

async function provisionPullRequestChannel(slackAPI: SlackAPIAdapter, payload: BitbucketNotification, usePrivateChannels: boolean) {
    return payload.eventKey == "pr:opened" ? await slackAPI.createChannel({
        name: buildChannelName(payload.pullRequest),
        isPrivate: usePrivateChannels
    }) : await slackAPI.findChannel(buildChannelName(payload.pullRequest), true);
}
