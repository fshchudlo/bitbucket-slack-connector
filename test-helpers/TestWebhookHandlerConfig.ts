import { WebhookHandlerConfig } from "../bitbucket-webhook-handler/webhookHandlerConfig";

export const TestWebhookHandlerConfig: WebhookHandlerConfig = {
    getOpenedPRBroadcastChannelId(): string | null {
        return null;
    },
    usePrivateChannels: true
};