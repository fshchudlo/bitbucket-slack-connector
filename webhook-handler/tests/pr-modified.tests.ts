import SlackAdapterSnapshottingMock from "./mocks/SlackAdapterSnapshottingMock";
import TestPayloadBuilder from "./mocks/TestPayloadBuilder";
import handleBitbucketWebhook from "../handleBitbucketWebhook";
import { TestBitbucketGateway } from "./mocks/TestBitbucketGateway";
import { BitbucketNotification } from "../../typings";
import { TestWebhookHandlerConfig } from "./mocks/TestWebhookHandlerConfig";

describe("handleBitbucketWebhook", () => {
    it("Should send message when PR modified", async () => {
        const testSlackGateway = new SlackAdapterSnapshottingMock();
        const payload = TestPayloadBuilder.pullRequestModified();
        await handleBitbucketWebhook(payload, testSlackGateway, new TestBitbucketGateway(), TestWebhookHandlerConfig);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });
    it("Doesn't send message if PR doesn't contain visible changes", async () => {
        const testSlackGateway = new SlackAdapterSnapshottingMock();
        const prCreatedPayload = TestPayloadBuilder.pullRequestOpened();
        const modifiedPayload = {
            ...prCreatedPayload,
            eventKey: "pr:modified",
            previousDescription: prCreatedPayload.pullRequest.description,
            previousTitle: prCreatedPayload.pullRequest.title,
            previousTarget: {
                displayId: prCreatedPayload.pullRequest.toRef.displayId,
                latestCommit: prCreatedPayload.pullRequest.toRef.latestCommit
            }
        } as BitbucketNotification;
        await handleBitbucketWebhook(modifiedPayload, testSlackGateway, new TestBitbucketGateway(), TestWebhookHandlerConfig);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });
});
