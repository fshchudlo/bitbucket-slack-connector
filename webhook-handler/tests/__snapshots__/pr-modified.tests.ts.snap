// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBitbucketWebhook Doesn't send message if PR doesn't contain visible changes 1`] = `
{
  "addedBookmarks": [],
  "addedReactions": [],
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [
    "pr-test-proj-test-repository-1",
  ],
  "searchedCommentSnapshots": [],
  "searchedPrOpenedBroadcastMessages": [],
  "sentMessages": [],
}
`;

exports[`handleBitbucketWebhook Should send message when PR modified 1`] = `
{
  "addedBookmarks": [],
  "addedReactions": [],
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [
    "pr-test-proj-test-repository-1",
  ],
  "searchedCommentSnapshots": [],
  "searchedPrOpenedBroadcastMessages": [],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": ":writing_hand: Test Author changed the pull request",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "Target is changed to \`not-the-master\`",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "type": "divider",
        },
        {
          "text": {
            "text": "Updated title:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "elements": [
            {
              "text": "New pull request title",
              "type": "mrkdwn",
            },
          ],
          "type": "context",
        },
        {
          "type": "divider",
        },
        {
          "text": {
            "text": "Updated description:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "elements": [
            {
              "text": "New pull request description",
              "type": "mrkdwn",
            },
          ],
          "type": "context",
        },
        {
          "type": "divider",
        },
        {
          "text": {
            "text": "Please <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|review the PR>",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": ":writing_hand: Test Author changed the pull request",
    },
  ],
}
`;
